File: backend/scripts/manage-users.js
#!/usr/bin/env node

const fs = require('fs');
const bcrypt = require('bcryptjs');
const readline = require('readline');

const HTPASSWD_FILE = '.htpasswd';

console.log('🔐 FIO Analyzer Admin User Management Tool');
console.log('==========================================\n');

// Hidden password input function
function getHiddenPassword(prompt) {
  return new Promise((resolve) => {
    process.stdout.write(prompt);

    // Check if stdin has setRawMode (TTY)
    if (process.stdin.setRawMode) {
      process.stdin.setRawMode(true);
      process.stdin.resume();
      process.stdin.setEncoding('utf8');

      let password = '';

      const onData = (char) => {
        switch (char) {
          case '\n':
          case '\r':
          case '\u0004': // Ctrl+D
            process.stdin.setRawMode(false);
            process.stdin.pause();
            process.stdin.removeListener('data', onData);
            process.stdout.write('\n');
            resolve(password);
            break;
          case '\u0003': // Ctrl+C
            process.stdout.write('\n');
            process.exit(1);
            break;
          case '\u007f': // Backspace
            if (password.length > 0) {
              password = password.slice(0, -1);
              process.stdout.write('\b \b');
            }
            break;
          default:
            password += char;
            process.stdout.write('*');
            break;
        }
      };

      process.stdin.on('data', onData);
    } else {
      // Fallback for non-TTY environments
      console.log('\n⚠️  Warning: Running in non-TTY mode - password will be visible!');
      console.log('💡 For secure password input, use: docker exec -it fio-app npm run change-password');
      console.log('💡 Or use non-interactive mode: node scripts/change-password.js username password\n');

      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      rl.question(prompt, (password) => {
        rl.close();
        resolve(password);
      });
    }
  });
}

async function main() {
  try {
    // Check for command line arguments
    const args = process.argv.slice(2);
    let username, password;

    if (args.length >= 2) {
      // Non-interactive mode: username and password from command line
      username = args[0];
      password = args[1];
      console.log(`👤 Username: ${username}`);
      console.log('🔑 Password provided via command line');
    } else {
      // Interactive mode
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      // Get username
      username = await new Promise((resolve) => {
        rl.question('Admin username (default: admin): ', (answer) => {
          resolve(answer.trim() || 'admin');
        });
      });

      rl.close();

      // Get password with hidden input
      password = await getHiddenPassword('New password: ');

      if (password && password.length >= 3) {
        const confirmPassword = await getHiddenPassword('Confirm password: ');

        if (password !== confirmPassword) {
          console.log('❌ Passwords do not match');
          process.exit(1);
        }
      }
    }

    if (!password || password.length < 3) {
      console.log('❌ Password must be at least 3 characters long');
      process.exit(1);
    }

    // Hash the password
    console.log('🔨 Hashing password...');
    const hashedPassword = await bcrypt.hash(password, 10);

    // Update .htpasswd file while preserving other users
    console.log('📝 Updating .htpasswd file...');

    let existingContent = '';
    if (fs.existsSync(HTPASSWD_FILE)) {
      existingContent = fs.readFileSync(HTPASSWD_FILE, 'utf8');
    }

    const lines = existingContent.split('\n').filter(line => line.trim() && !line.startsWith('#'));
    const userExists = lines.findIndex(line => line.startsWith(`${username}:`));

    if (userExists !== -1) {
      // Update existing user
      lines[userExists] = `${username}:${hashedPassword}`;
      console.log(`👤 Updated existing user '${username}'`);
    } else {
      // Add new user
      lines.push(`${username}:${hashedPassword}`);
      console.log(`👤 Added new user '${username}'`);
    }

    // Write back to file with preserved structure
    const newContent = lines.join('\n') + '\n';
    fs.writeFileSync(HTPASSWD_FILE, newContent, 'utf8');

    console.log('✅ Admin user updated successfully!');
    console.log(`👤 Username: ${username}`);
    console.log('📁 Role: Administrator (full access to all features)');
    console.log('🔄 Restart the application to apply changes.');

  } catch (error) {
    console.error('❌ Error changing password:', error.message);
    process.exit(1);
  }
}

// Usage information
if (process.argv.includes('--help') || process.argv.includes('-h')) {
  console.log('Usage:');
  console.log('  Interactive mode:     docker exec -it fio-app npm run manage-users');
  console.log('  Non-interactive mode: docker exec fio-app node scripts/manage-users.js username password');
  console.log('  Help:                 npm run manage-users --help');
  console.log('');
  console.log('Examples:');
  console.log('  docker exec fio-app node scripts/manage-users.js admin newpassword');
  console.log('  docker exec fio-app node scripts/manage-users.js alice secretpass');
  console.log('');
  console.log('Note: Admin users have full access to view, edit, delete, and upload data.');
  process.exit(0);
}

// Check if running in Docker container
if (!fs.existsSync('/app')) {
  console.error('❌ This script should be run inside the Docker container');
  console.error('💡 Use: docker exec -it fio-app npm run manage-users');
  console.error('💡 Or:   docker exec fio-app node scripts/manage-users.js username password');
  process.exit(1);
}

main();

File: backend/scripts/manage_users.py
#!/usr/bin/env python3
"""
User management script for FIO Analyzer
"""

import argparse
import sys
from pathlib import Path

import bcrypt

# Add parent directory to path to import modules
sys.path.insert(0, str(Path(__file__).parent.parent))

from auth.authentication import parse_htpasswd
from config.settings import settings


def hash_password(password: str) -> str:
    """Hash password using bcrypt"""
    return bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")


def add_user(username: str, password: str, file_path: Path):
    """Add user to htpasswd file"""
    hashed_password = hash_password(password)

    # Read existing users
    users = parse_htpasswd(file_path) or {}

    # Add new user
    users[username] = hashed_password

    # Write back to file
    with open(file_path, "w") as f:
        for user, pwd_hash in users.items():
            f.write(f"{user}:{pwd_hash}\n")

    print(f"User '{username}' added successfully to {file_path}")


def remove_user(username: str, file_path: Path):
    """Remove user from htpasswd file"""
    users = parse_htpasswd(file_path) or {}

    if username not in users:
        print(f"User '{username}' not found in {file_path}")
        return

    del users[username]

    # Write back to file
    with open(file_path, "w") as f:
        for user, pwd_hash in users.items():
            f.write(f"{user}:{pwd_hash}\n")

    print(f"User '{username}' removed successfully from {file_path}")


def list_users(file_path: Path):
    """List all users in htpasswd file"""
    users = parse_htpasswd(file_path) or {}

    if not users:
        print(f"No users found in {file_path}")
        return

    print(f"Users in {file_path}:")
    for username in users.keys():
        print(f"  - {username}")


def main():
    parser = argparse.ArgumentParser(description="Manage FIO Analyzer users")
    parser.add_argument("action", choices=["add", "remove", "list"], help="Action to perform")
    parser.add_argument("--username", "-u", help="Username")
    parser.add_argument("--password", "-p", help="Password")
    parser.add_argument(
        "--uploader",
        action="store_true",
        help="Manage uploader users (default: admin users)",
    )

    args = parser.parse_args()

    # Determine which file to use
    file_path = settings.htuploaders_path if args.uploader else settings.htpasswd_path

    if args.action == "add":
        if not args.username or not args.password:
            print("Username and password are required for adding users")
            sys.exit(1)
        add_user(args.username, args.password, file_path)

    elif args.action == "remove":
        if not args.username:
            print("Username is required for removing users")
            sys.exit(1)
        remove_user(args.username, file_path)

    elif args.action == "list":
        list_users(file_path)


if __name__ == "__main__":
    main()
